test/boost/mutation_reader_test.cc

```diff
@@ -3520,3 +3521,139 @@ SEASTAR_THREAD_TEST_CASE(test_evictable_reader_self_validation) {
+        for (auto k: all_ks) {
+            auto ck = mk_ck(k);
+            positions.push_back(position_in_partition::before_key(ck));
+            positions.push_back(position_in_partition::for_key(ck));
+            positions.push_back(position_in_partition::after_key(ck));
+        }
+        positions.push_back(position_in_partition::after_all_clustered_rows());
+
+        auto num_ranges = tests::random::get_int(1ul, std::max(all_ks.size() / 3, 1ul));
+        positions = tests::random::random_subset(std::move(positions), num_ranges * 2, engine);
+        std::sort(positions.begin(), positions.end(), position_in_partition::less_compare(*_s));
+
+        std::vector<position_range> fwd_ranges;
+        for (int i = 0; i < num_ranges; ++i) {
+            assert(2*i+1 < positions.size());
+            fwd_ranges.push_back(position_range(std::move(positions[2*i]), std::move(positions[2*i+1])));
+        }
+
+        return scenario {
+            std::move(readers_data),
+            std::move(fwd_ranges)
+        };
+    }
+};
+
+SEASTAR_THREAD_TEST_CASE(test_clustering_order_merger_in_memory) {
+    clustering_order_merger_test_generator g;
+
+    auto make_authority = [] (mutation mut, streamed_mutation::forwarding fwd) {
+        return flat_mutation_reader_from_mutations(tests::make_permit(), {std::move(mut)}, fwd);
+    };
+
+    auto make_tested = [s = g._s] (std::vector<mutation_bounds> ms, streamed_mutation::forwarding fwd) {
+        auto rs = boost::copy_range<std::vector<reader_bounds>>(std::move(ms)
+                | boost::adaptors::transformed([fwd] (auto&& mb) {
+                    return reader_bounds{
+                        flat_mutation_reader_from_mutations(tests::make_permit(), {std::move(mb.m)}, fwd),
+                        std::move(mb.lower), std::move(mb.upper)};
+                }));
+        auto q = std::make_unique<simple_position_reader_queue>(*s, std::move(rs));
+        return make_clustering_combined_reader(s, tests::make_permit(), fwd, std::move(q));
+    };
+
+    auto seed = tests::random::get_int<uint32_t>();
+    std::cout << "test_clustering_order_merger_in_memory seed: " << seed << std::endl;
+    auto engine = std::mt19937(seed);
+
+    for (int run = 0; run < 1000; ++run) {
+        auto scenario = g.generate_scenario(engine);
+        auto merged = std::accumulate(scenario.readers_data.begin(), scenario.readers_data.end(),
+                mutation(g._s, g._pk), [] (mutation curr, const mutation_bounds& mb) { return std::move(curr) + mb.m; });
+
+        {
+            auto fwd = streamed_mutation::forwarding::no;
+            compare_readers(*g._s, make_authority(merged, fwd), make_tested(scenario.readers_data, fwd));
+        }
+
+        auto fwd = streamed_mutation::forwarding::yes;
+        compare_readers(*g._s, make_authority(std::move(merged), fwd),
+                make_tested(std::move(scenario.readers_data), fwd), scenario.fwd_ranges);
+    }
+}
```

Comment:
I would also like to see a test with `run_mutation_source_tests()`. To work-around the single partition limitation, you can write a meta reader which constructs a new merging reader on every new partition.
For any mutation that has a static row or partition tombstone you can just bail-out and create `flat_mutation_reader_from_mutations()` instead. For increased effects you can de-compose mutations having many rows into many mutations containing non-overlapping clustering ranges.

I cannot count the amount of times the fast-forwarding stress tests in this suite shaken out hidden bugs from my readers that the often simplistic tests I wrote didn't catch.